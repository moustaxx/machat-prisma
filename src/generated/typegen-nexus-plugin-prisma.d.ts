import * as Typegen from 'nexus-plugin-prisma/typegen'
import * as Prisma from '@prisma/client';

// Pagination type
type Pagination = {
  first?: boolean
  last?: boolean
  before?: boolean
  after?: boolean
}

// Prisma custom scalar names
type CustomScalars = 'DateTime' | 'Json'

// Prisma model type definitions
interface PrismaModels {
  Conversation: Prisma.Conversation
  Message: Prisma.Message
  Person: Prisma.Person
  LastRead: Prisma.LastRead
  Session: Prisma.Session
}

// Prisma input types metadata
interface NexusPrismaInputs {
  Query: {
    conversations: {
      filtering: 'AND' | 'OR' | 'NOT' | 'createdAt' | 'id' | 'name' | 'messages' | 'lastRead' | 'participants'
      ordering: 'createdAt' | 'id' | 'name'
    }
    messages: {
      filtering: 'AND' | 'OR' | 'NOT' | 'content' | 'authorID' | 'conversationID' | 'createdAt' | 'id' | 'author' | 'conversation'
      ordering: 'content' | 'authorID' | 'conversationID' | 'createdAt' | 'id'
    }
    people: {
      filtering: 'AND' | 'OR' | 'NOT' | 'createdAt' | 'email' | 'id' | 'isAdmin' | 'lastSeen' | 'username' | 'hash' | 'messages' | 'lastRead' | 'conversations'
      ordering: 'createdAt' | 'email' | 'id' | 'isAdmin' | 'lastSeen' | 'username' | 'hash'
    }
    lastReads: {
      filtering: 'AND' | 'OR' | 'NOT' | 'lastRead' | 'personID' | 'conversationID' | 'person' | 'conversation'
      ordering: 'lastRead' | 'personID' | 'conversationID'
    }
    sessions: {
      filtering: 'AND' | 'OR' | 'NOT' | 'sid' | 'sess' | 'expire'
      ordering: 'sid' | 'sess' | 'expire'
    }
  },
  Conversation: {
    messages: {
      filtering: 'AND' | 'OR' | 'NOT' | 'content' | 'authorID' | 'conversationID' | 'createdAt' | 'id' | 'author' | 'conversation'
      ordering: 'content' | 'authorID' | 'conversationID' | 'createdAt' | 'id'
    }
    lastRead: {
      filtering: 'AND' | 'OR' | 'NOT' | 'lastRead' | 'personID' | 'conversationID' | 'person' | 'conversation'
      ordering: 'lastRead' | 'personID' | 'conversationID'
    }
    participants: {
      filtering: 'AND' | 'OR' | 'NOT' | 'createdAt' | 'email' | 'id' | 'isAdmin' | 'lastSeen' | 'username' | 'hash' | 'messages' | 'lastRead' | 'conversations'
      ordering: 'createdAt' | 'email' | 'id' | 'isAdmin' | 'lastSeen' | 'username' | 'hash'
    }
  }
  Message: {

  }
  Person: {
    messages: {
      filtering: 'AND' | 'OR' | 'NOT' | 'content' | 'authorID' | 'conversationID' | 'createdAt' | 'id' | 'author' | 'conversation'
      ordering: 'content' | 'authorID' | 'conversationID' | 'createdAt' | 'id'
    }
    lastRead: {
      filtering: 'AND' | 'OR' | 'NOT' | 'lastRead' | 'personID' | 'conversationID' | 'person' | 'conversation'
      ordering: 'lastRead' | 'personID' | 'conversationID'
    }
    conversations: {
      filtering: 'AND' | 'OR' | 'NOT' | 'createdAt' | 'id' | 'name' | 'messages' | 'lastRead' | 'participants'
      ordering: 'createdAt' | 'id' | 'name'
    }
  }
  LastRead: {

  }
  Session: {

  }
}

// Prisma output types metadata
interface NexusPrismaOutputs {
  Query: {
    conversation: 'Conversation'
    conversations: 'Conversation'
    message: 'Message'
    messages: 'Message'
    person: 'Person'
    people: 'Person'
    lastRead: 'LastRead'
    lastReads: 'LastRead'
    session: 'Session'
    sessions: 'Session'
  },
  Mutation: {
    createOneConversation: 'Conversation'
    updateOneConversation: 'Conversation'
    updateManyConversation: 'BatchPayload'
    deleteOneConversation: 'Conversation'
    deleteManyConversation: 'BatchPayload'
    upsertOneConversation: 'Conversation'
    createOneMessage: 'Message'
    updateOneMessage: 'Message'
    updateManyMessage: 'BatchPayload'
    deleteOneMessage: 'Message'
    deleteManyMessage: 'BatchPayload'
    upsertOneMessage: 'Message'
    createOnePerson: 'Person'
    updateOnePerson: 'Person'
    updateManyPerson: 'BatchPayload'
    deleteOnePerson: 'Person'
    deleteManyPerson: 'BatchPayload'
    upsertOnePerson: 'Person'
    createOneLastRead: 'LastRead'
    updateOneLastRead: 'LastRead'
    updateManyLastRead: 'BatchPayload'
    deleteOneLastRead: 'LastRead'
    deleteManyLastRead: 'BatchPayload'
    upsertOneLastRead: 'LastRead'
    createOneSession: 'Session'
    updateOneSession: 'Session'
    updateManySession: 'BatchPayload'
    deleteOneSession: 'Session'
    deleteManySession: 'BatchPayload'
    upsertOneSession: 'Session'
  },
  Conversation: {
    createdAt: 'DateTime'
    id: 'Int'
    name: 'String'
    messages: 'Message'
    lastRead: 'LastRead'
    participants: 'Person'
  }
  Message: {
    content: 'String'
    authorID: 'Int'
    conversationID: 'Int'
    createdAt: 'DateTime'
    id: 'Int'
    author: 'Person'
    conversation: 'Conversation'
  }
  Person: {
    createdAt: 'DateTime'
    email: 'String'
    id: 'Int'
    isAdmin: 'Boolean'
    lastSeen: 'DateTime'
    username: 'String'
    hash: 'String'
    messages: 'Message'
    lastRead: 'LastRead'
    conversations: 'Conversation'
  }
  LastRead: {
    lastRead: 'DateTime'
    personID: 'Int'
    conversationID: 'Int'
    person: 'Person'
    conversation: 'Conversation'
  }
  Session: {
    sid: 'String'
    sess: 'Json'
    expire: 'DateTime'
  }
}

// Helper to gather all methods relative to a model
interface NexusPrismaMethods {
  Conversation: Typegen.NexusPrismaFields<'Conversation'>
  Message: Typegen.NexusPrismaFields<'Message'>
  Person: Typegen.NexusPrismaFields<'Person'>
  LastRead: Typegen.NexusPrismaFields<'LastRead'>
  Session: Typegen.NexusPrismaFields<'Session'>
  Query: Typegen.NexusPrismaFields<'Query'>
  Mutation: Typegen.NexusPrismaFields<'Mutation'>
}

interface NexusPrismaGenTypes {
  inputs: NexusPrismaInputs
  outputs: NexusPrismaOutputs
  methods: NexusPrismaMethods
  models: PrismaModels
  pagination: Pagination
  scalars: CustomScalars
}

declare global {
  interface NexusPrismaGen extends NexusPrismaGenTypes {}

  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = Typegen.GetNexusPrisma<TypeName, ModelOrCrud>;
}
  