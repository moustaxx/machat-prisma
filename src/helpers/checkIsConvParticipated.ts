import { ForbiddenError } from 'apollo-server-errors';

import { Context } from '../context';

const checkIsConvParticipated = async (
    prisma: Context['prisma'],
    sessionOwner: NonNullable<Context['session']>['owner'],
    conversationId: number,
): Promise<void> => {
    const getConv = await prisma.person.findOne({
        select: {
            conversations: {
                select: {
                    id: true,
                },
            },
        },
        where: { id: sessionOwner?.id },
    });

    const isParticipated = !!getConv?.conversations.find((conv) => {
        return conv.id === conversationId;
    });

    if (!sessionOwner?.isAdmin && !isParticipated) {
        throw new ForbiddenError('Insufficient permissions');
    }
};

export default checkIsConvParticipated;
