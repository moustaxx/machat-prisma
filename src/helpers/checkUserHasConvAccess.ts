import { ForbiddenError } from 'apollo-server-errors';
import { NexusGenAllTypes } from '../generated/nexus';
import { TPrisma } from '../prismaClient';

const checkUserHasConvAccess = async (
    prisma: TPrisma,
    user: NexusGenAllTypes['Person'],
    conversationId: number,
): Promise<void> => {
    const getConv = await prisma.person.findOne({
        select: {
            conversations: {
                select: {
                    id: true,
                },
            },
        },
        where: { id: user?.id },
    });

    const isParticipated = !!getConv?.conversations.find((conv) => {
        return conv.id === conversationId;
    });

    if (!user?.isAdmin && !isParticipated) {
        throw new ForbiddenError('Insufficient permissions');
    }
};

export default checkUserHasConvAccess;
