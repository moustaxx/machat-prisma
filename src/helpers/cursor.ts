type Model = {
    id?: number;
};
type Cursor = {
    id: number;
};

const cursorUtils = {
    getCursor: ({ id }: Model): Cursor => {
        if (!id) throw new Error('Model doesn\'t have an ID');
        return { id };
    },
    encodeCursor: ({ id }: Model): string => {
        if (!id) throw new Error('Model doesn\'t have an ID');
        return Buffer.from(JSON.stringify({ id })).toString('base64');
    },
    decodeCursor: (cursor: string): Cursor => {
        return JSON.parse(Buffer.from(cursor, 'base64').toString('ascii')) as Cursor;
    },
};

export default cursorUtils;
