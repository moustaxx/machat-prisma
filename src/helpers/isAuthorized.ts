import { ApolloError } from 'apollo-server-errors';
import { ISession } from '../types';

function isAuthorized(session?: ISession): asserts session is
ISession & Pick<Required<ISession>, 'owner'> {
    if (!session || !session.isLoggedIn || !session.owner) {
        throw new ApolloError('You must be logged in!', 'UNAUTHORIZED');
    }
}

export default isAuthorized;
