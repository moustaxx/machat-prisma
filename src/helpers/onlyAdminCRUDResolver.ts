import { CustomFieldResolver } from 'nexus-plugin-prisma/typegen';
import { ForbiddenError } from 'apollo-server-errors';

const onlyAdminCRUDResolver: CustomFieldResolver<any, any> = (
    root,
    args,
    ctx,
    info,
    originalResolve,
) => {
    if (!ctx.session?.owner?.isAdmin) {
        throw new ForbiddenError('Insufficient permissions');
    }
    return originalResolve(root, args, ctx, info);
};

export default onlyAdminCRUDResolver;
