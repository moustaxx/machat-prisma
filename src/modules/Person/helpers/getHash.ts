import { scryptSync } from 'crypto';

export const getHash = (password: string, salt: string): string => {
    return scryptSync(password, salt, 32).toString('hex');
};
