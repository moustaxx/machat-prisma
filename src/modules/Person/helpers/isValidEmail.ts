const regexp = /^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;

const isValidEmail = (email: string): boolean => {
    if (!email || email.length > 64) return false;
    return regexp.test(email);
};

export default isValidEmail;
