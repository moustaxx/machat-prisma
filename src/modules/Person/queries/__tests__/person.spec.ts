import { PrismaClient } from '@prisma/client';
import { FastifyInstance } from 'fastify';
import { createFastifyGQLTestClient, GQLResponse } from 'fastify-gql-integration-testing';

import main from '../../../..';
import { NexusGenRootTypes } from '../../../../generated/nexus';
import { gqlRequest, createRandomUserAndLogin } from '../../../../tests/helpers';

let client: PrismaClient;
let app: FastifyInstance;
let testClient: ReturnType<typeof createFastifyGQLTestClient>;

beforeAll(async () => {
    client = new PrismaClient();
    app = await main(true);
    testClient = createFastifyGQLTestClient(app);
});

afterAll(async () => {
    await Promise.allSettled([
        client.$disconnect(),
        app.close(),
    ]);
});

it('should throw FORBIDDEN error when quering people without permissions', async () => {
    const data = await testClient.query(`
        query getPeople {
            people {
                id
            }
        }
    `);

    const errorCode = data.errors && data.errors[0].extensions?.code;
    expect(errorCode).toEqual('FORBIDDEN');
});

it('should return people list when admin permissions are present d', async () => {
    const { cookies } = await createRandomUserAndLogin(app, client, { isAdmin: true });

    const peopleRes = await gqlRequest(app, {
        cookies,
        query: `
            query getPeople {
                people {
                    id
                }
            }
        `,
    });

    type TPeopleQuery = GQLResponse<{ people: NexusGenRootTypes['Person'][] }>;
    const { data }: TPeopleQuery = await peopleRes.json();

    expect(data.people.length).toBeGreaterThan(0);
});
