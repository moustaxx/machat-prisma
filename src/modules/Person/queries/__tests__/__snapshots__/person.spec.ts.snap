// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return person 1`] = `
Object {
  "person": Object {
    "conversations": Array [],
    "createdAt": Any<String>,
    "email": Any<String>,
    "id": 2,
    "isActive": true,
    "isAdmin": null,
    "lastSeen": null,
    "messages": Array [],
    "username": Any<String>,
  },
}
`;
